File: packages/engineering-validation/requirements/ambiguity.ts
/**
 * Detects qualitative and non-deterministic requirement terms.
 */

export class AmbiguityDetector {
  private static readonly VAGUE_TERMS = [
    "low power",
    "ultra low power",
    "compact",
    "small form factor",
    "fast",
    "high speed",
    "efficient",
    "cheap",
    "low cost",
    "heavy duty",
    "rugged",
    "lightweight",
  ];

  public static checkAmbiguity(text: string): { isAmbiguous: boolean; detectedTerms: string[] } {
    const lower = text.toLowerCase();
    const detected: string[] = [];

    for (const term of this.VAGUE_TERMS) {
      if (lower.includes(term)) {
        detected.push(term);
      }
    }

    return {
      isAmbiguous: detected.length > 0,
      detectedTerms: detected,
    };
  }
}

File: backend/workline/validation/ambiguity.py
"""Detects qualitative, ambiguous requirement terms."""

from typing import Dict, List


class AmbiguityDetector:
    VAGUE_TERMS = [
        "low power",
        "ultra low power",
        "compact",
        "small form factor",
        "fast",
        "high speed",
        "efficient",
        "cheap",
        "low cost",
        "heavy duty",
        "rugged",
        "lightweight",
    ]

    @classmethod
    def check_ambiguity(cls, text: str) -> Dict[str, any]:
        lower = text.lower()
        detected = [term for term in cls.VAGUE_TERMS if term in lower]
        return {
            "is_ambiguous": len(detected) > 0,
            "isAmbiguous": len(detected) > 0,
            "detected_terms": detected,
            "detectedTerms": detected,
        }

    @classmethod
    def checkAmbiguity(cls, text: str) -> Dict[str, any]:
        return cls.check_ambiguity(text)
